README.md
# PHYExtendedAppDelegate

[![Version](http://cocoapod-badges.herokuapp.com/v/PHYExtendedAppDelegate/badge.png)](http://cocoadocs.org/docsets/PHYExtendedAppDelegate)
[![Platform](http://cocoapod-badges.herokuapp.com/p/PHYExtendedAppDelegate/badge.png)](http://cocoadocs.org/docsets/PHYExtendedAppDelegate)

## Usage

To run the example project; clone the repo, and run `pod install` from the Example directory first.

## Requirements

## Installation

PHYExtendedAppDelegate is available through [CocoaPods](http://cocoapods.org), to install
it simply add the following line to your Podfile:

    pod "PHYExtendedAppDelegate"

## Author

Matt Ricketson, mricketson@mac.com

## License

PHYExtendedAppDelegate is available under the MIT license. See the LICENSE file for more info.



CHANGELOG.md
# PHYExtendedAppDelegate CHANGELOG

## 0.1.0

Initial release.


Classes/PHYExtendedAppDelegate.h
//
//  PHStandardAppDelegate.h
//  Rally
//
//  Created by Matt Ricketson on 11/25/13.
//  Copyright (c) 2013 Phyre. All rights reserved.
//

#import <UIKit/UIKit.h>

extern NSString * const PHYAppDelegateDidOpenURLNotification;
extern NSString * const PHYAppDelegateDidOpenURLSourceApplicationKey;
extern NSString * const PHYAppDelegateDidOpenURLAnnotationKey;
extern NSString * const PHYAppDelegateDidOpenURLURLKey;
extern NSString * const PHYAppDelegateDidOpenURLQueryDataKey;

extern BOOL phy_isRunningTests(void) __attribute__((const));

typedef void(^PHYRemoteNotificationRegistrationSuccesshandler)(NSData *deviceToken);
typedef void(^PHYRemoteNotificationRegistrationFailurehandler)(NSError *error);



#pragma mark -

/**
 `PHAppDelegate` is an app delegate superclass designed to improve the convenience of common app delegate tasks such as remote notification registration.
 */
@interface PHYExtendedAppDelegate : UIResponder <UIApplicationDelegate>

///----------------------
/// @name Shared Instance
///----------------------

/// The delegate for the current application.
+ (instancetype)sharedDelegate;


///--------------------------
/// @name App Characteristics
///--------------------------

/// The application window.
@property (nonatomic, strong) UIWindow *window;


///------------------------------------
/// @name Handling Remote Notifications
///------------------------------------

- (void)registerForRemoteNotifications:(PHYRemoteNotificationRegistrationSuccesshandler)success
							   failure:(PHYRemoteNotificationRegistrationFailurehandler)failure;

- (void)registerForRemoteNotificationTypes:(UIRemoteNotificationType)types
								   success:(PHYRemoteNotificationRegistrationSuccesshandler)success
								   failure:(PHYRemoteNotificationRegistrationFailurehandler)failure;


///-------------------------
/// @name Handling Open URLs
///-------------------------

- (NSDictionary *)dictionaryForParsedQueryString:(NSString *)query;

@end



#pragma mark - UIApplication Extensions

extern NSString * NSStringFromUIApplicationState(UIApplicationState appState);


Example/PHYExtendedAppDelegate/PHYAppDelegate.m
//
//  PHYAppDelegate.m
//  PHYExtendedAppDelegate
//
//  Created by Matt Ricketson on 4/9/14.
//  Copyright (c) 2014 Phyre Inc. All rights reserved.
//

#import "PHYAppDelegate.h"

@implementation PHYAppDelegate

- (BOOL)application:(UIApplication *)application didFinishLaunchingWithOptions:(NSDictionary *)launchOptions
{
    // Override point for customization after application launch.
    return YES;
}
							
- (void)applicationWillResignActive:(UIApplication *)application
{
	// Sent when the application is about to move from active to inactive state. This can occur for certain types of temporary interruptions (such as an incoming phone call or SMS message) or when the user quits the application and it begins the transition to the background state.
	// Use this method to pause ongoing tasks, disable timers, and throttle down OpenGL ES frame rates. Games should use this method to pause the game.
}

- (void)applicationDidEnterBackground:(UIApplication *)application
{
	// Use this method to release shared resources, save user data, invalidate timers, and store enough application state information to restore your application to its current state in case it is terminated later. 
	// If your application supports background execution, this method is called instead of applicationWillTerminate: when the user quits.
}

- (void)applicationWillEnterForeground:(UIApplication *)application
{
	// Called as part of the transition from the background to the inactive state; here you can undo many of the changes made on entering the background.
}

- (void)applicationDidBecomeActive:(UIApplication *)application
{
	// Restart any tasks that were paused (or not yet started) while the application was inactive. If the application was previously in the background, optionally refresh the user interface.
}

- (void)applicationWillTerminate:(UIApplication *)application
{
	// Called when the application is about to terminate. Save data if appropriate. See also applicationDidEnterBackground:.
}

@end


Classes/PHYExtendedAppDelegate.m
//
//  PHStandardAppDelegate.m
//  Rally
//
//  Created by Matt Ricketson on 11/25/13.
//  Copyright (c) 2013 Phyre. All rights reserved.
//

#import "PHYExtendedAppDelegate.h"

NSString * const PHYAppDelegateDidOpenURLNotification			= @"PHYAppDelegateDidOpenURLNotification";
NSString * const PHYAppDelegateDidOpenURLSourceApplicationKey	= @"PHYAppDelegateDidOpenURLSourceApplicationKey";
NSString * const PHYAppDelegateDidOpenURLAnnotationKey			= @"PHYAppDelegateDidOpenURLAnnotationKey";
NSString * const PHYAppDelegateDidOpenURLURLKey					= @"PHYAppDelegateDidOpenURLURLKey";
NSString * const PHYAppDelegateDidOpenURLQueryDataKey			= @"PHYAppDelegateDidOpenURLQueryDataKey";

BOOL phy_isRunningTests(void)
{
	NSDictionary *environment = [[NSProcessInfo processInfo] environment];
    NSString *injectBundle = environment[@"XCInjectBundle"];
	NSString *injectBundlePathExtension = [injectBundle pathExtension];
    return [injectBundlePathExtension isEqualToString:@"xctest"];
}


@interface PHYExtendedAppDelegate ()

@property (nonatomic, copy) PHYRemoteNotificationRegistrationSuccesshandler remoteNotificationRegistrationSuccessHandler;
@property (nonatomic, copy) PHYRemoteNotificationRegistrationFailurehandler remoteNotificationRegistrationFailureHandler;

@end


@implementation PHYExtendedAppDelegate

#pragma mark - Shared Instance

+ (instancetype)sharedDelegate
{
	return [[UIApplication sharedApplication] delegate];
}


#pragma mark - Remote Notifications

- (void)registerForRemoteNotifications:(PHYRemoteNotificationRegistrationSuccesshandler)success
							   failure:(PHYRemoteNotificationRegistrationFailurehandler)failure
{
	[self registerForRemoteNotificationTypes:(UIRemoteNotificationTypeBadge | UIRemoteNotificationTypeSound | UIRemoteNotificationTypeAlert)
									 success:success
									 failure:failure];
}

- (void)registerForRemoteNotificationTypes:(UIRemoteNotificationType)types
								   success:(PHYRemoteNotificationRegistrationSuccesshandler)success
								   failure:(PHYRemoteNotificationRegistrationFailurehandler)failure
{
	self.remoteNotificationRegistrationSuccessHandler = success;
	self.remoteNotificationRegistrationFailureHandler = failure;
	
	[[UIApplication sharedApplication] registerForRemoteNotificationTypes:types];
}

- (void)application:(UIApplication *)application didRegisterForRemoteNotificationsWithDeviceToken:(NSData *)deviceToken
{
	if (self.remoteNotificationRegistrationSuccessHandler) {
		self.remoteNotificationRegistrationSuccessHandler(deviceToken);
	}
	
	self.remoteNotificationRegistrationSuccessHandler = nil;
	self.remoteNotificationRegistrationFailureHandler = nil;
}

- (void)application:(UIApplication *)application didFailToRegisterForRemoteNotificationsWithError:(NSError *)error
{
	if (self.remoteNotificationRegistrationFailureHandler) {
		self.remoteNotificationRegistrationFailureHandler(error);
	}
	
	self.remoteNotificationRegistrationSuccessHandler = nil;
	self.remoteNotificationRegistrationFailureHandler = nil;
}


#pragma mark - Open URLs

- (BOOL)application:(UIApplication *)application openURL:(NSURL *)url sourceApplication:(NSString *)sourceApplication annotation:(id)annotation
{
	NSDictionary *userInfo = @{
							   PHYAppDelegateDidOpenURLSourceApplicationKey:	sourceApplication,
							   PHYAppDelegateDidOpenURLAnnotationKey:			annotation,
							   PHYAppDelegateDidOpenURLURLKey:					url,
							   PHYAppDelegateDidOpenURLQueryDataKey:			[self dictionaryForParsedQueryString:[url query]]
							   };
	
	[[NSNotificationCenter defaultCenter] postNotificationName:PHYAppDelegateDidOpenURLNotification object:self userInfo:userInfo];
	
	return YES;
}

- (NSDictionary *)dictionaryForParsedQueryString:(NSString *)query
{
    NSMutableDictionary *queryData = [NSMutableDictionary dictionary];
    NSArray *pairs = [query componentsSeparatedByString:@"&"];
    
    for (NSString *pair in pairs) {
        NSArray *elements = [pair componentsSeparatedByString:@"="];
        NSString *key = [elements[0] stringByReplacingPercentEscapesUsingEncoding:NSUTF8StringEncoding];
        NSString *val = [elements[1] stringByReplacingPercentEscapesUsingEncoding:NSUTF8StringEncoding];
        queryData[key] = val;
    }
	
    return queryData;
}

@end



#pragma mark - UIApplication Extensions

NSString * NSStringFromUIApplicationState(UIApplicationState appState)
{
	switch (appState) {
		case UIApplicationStateActive:		return @"UIApplicationStateActive";
		case UIApplicationStateInactive:	return @"UIApplicationStateInactive";
		case UIApplicationStateBackground:	return @"UIApplicationStateBackground";
	}
}
